### ISSUE TITLE
SC2012 wiki should cover more cases

### ISSUE THREAD

**Issue by Jemi** [koalaman/shellcheck#1852]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
SC2012
- My shellcheck version (`shellcheck --version` or "online"): 
Any
- [*] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2012)

Specifically, SC2012 wiki should cover the case where ls is being used just to get the number of files matching a certain literal or glob pattern, rather than propagating the actual name of any particular file outside of that line of code.

example:
`NUMGZ="$(ls -l *.gz | wc -l)"`

Either this is a viable use case and it should go in the exceptions (or even detected from within shellcheck), or it is not and the examples should explain what's wrong with it.

**Comment by dimo414** [comment]
You don't need to use `ls` to trigger glob expansion. Any unquoted string is subject to globbing regardless of context. If you want to expand a glob and count the number of matches an array is probably easiest:

```
gz_files=(*.gz)
numgz=${#gz_files[@]}
```

I don't think there should be an expectation that SC pages must _exhaustively_ list all possible circumstances. Just because your particular use case isn't mentioned doesn't mean it's a false positive.

**Comment by matthewpersico** [comment]
Pull requests accepted. 😉 Actually, not even: the Wiki is directly community editable. So, feel free to contribute by adding your desired example to the appropriate Wiki page.

**Comment by Jemi** [comment]
I don't really know shell well enough to justify contributing to the wiki, and my contribution would just be a question. I'm pretty much learning shell by using shellcheck (although I've used shell for more than 1 decade - I never really started to get it until I found shellcheck). 

My reason for posting this here as a bug, is so many shellcheck codes are listed with 'exceptions: none' that I naturally assume that shellcheck's coverage is intended to be exhaustive and definitive. And here I find a case where I can't determine whether my pattern is an acceptable use, or if there's something wrong with it that I just can't figure out. I don't know how ls behaves (is the number of lines supposed to be consistent under certain assumptions) and I don't even know the answer to that question when it comes to globs (I mean filenames can have spaces and newlines after all...). So if shellcheck doesn't give me the answer, I'll have to strain my eyes reading posix documentation, and then probably forget the answer the next day. So I do hope that the wiki can be as exhaustive as I had assumed it was supposed to be.

**Comment by matthewpersico** [comment]
I decided to add getting the *count* of filenames to the docs of getting a listing of the filenames. On the surface it seems like a corner case, just a different manipulation of a list, but then @dimo414 came along and gave an example that introduces arrays to the reader. Bash arrays are a construct, which, to my experience, is vastly underused (because it is under-known?), only exceeded by associative arrays in underusedness. (Yeah, that's probably not a word, but you get my meaning, so I'm using it.)  Many algorithms are more easily handled with real arrays and teaching more people how to use them may allow someone to stay in bash, as opposed to being forced to "move up" to Python or Perl or some other higher level language.

So, I went and looked at the page in question and the examples looked suspiciously like a dir listing of my home directory at work. Sure enough, guess who's ugly mug showed on the edit history of the page" 🤣 . I updated the page.

**Comment by Jemi** [comment]
I appreciate the update. Thanks.

**Comment by esthon** [comment]
gz_files=(*.gz)
numgz=${#gz_files[@]}

This does not work if there are no .gz files... $numgz will erroneously say that there are 1 instead of 0.

**Comment by dimo414** [comment]
See the `nullglob` options if no matches are expected, as well as [this discussion](https://unix.stackexchange.com/q/204803/19157) of why it's not the default.

**Comment by rdebath** [comment]
`FN=$(ls -tr data/logs/*.log | tail -1)`

This gives SC2012, notice the required sort order, is it a false positive?
There appears to be no viable alternative ... ?

**Comment by dimo414** [comment]
A search for "[bash find most recently modified file](https://www.google.com/search?q=bash+find+most+recently+modified+file)" finds [several suggestions using `find`](https://askubuntu.com/q/61179/12220), do any of those help?

**Comment by rdebath** [comment]
Those "several suggestions" seem to have the same filename characteristics as `ls -tr` which BTW detects that it's not writing to a tty and automatically switches to a plain single column format. 

In addition they use an intermediate text file to sort the files by date which may break permanently if a file with a newline exists in the directory. The `ls -tr` actually sorts by the date, so once the logfile gets written to again any junk file is ignored because it's older.

I suppose I could limit it to GNU tools and use the `-z` options to make those suggestions work; but if I do that I might as well use `-q` on `ls` because nuls and newlines are not going to be in valid filenames anyway.

**Comment by kankaristo** [comment]
Is the [rationale for SC2012](https://github.com/koalaman/shellcheck/wiki/SC2012) relevant when using a "plain" `ls` instead of `ls -l`? I'm genuinely asking, because the current rationale really only explains problematic usage of `ls -l`.

The current rationale also doesn't give the real filenames for the three "seemingly identical" filenames, nor does it show the equivalent output for `find` for those filenames. So, I'm not sure what kind of filenames would be problematic with `ls` vs. `find`.

The other examples for replacing `ls` with `find` don't actually seem to contain any really problematic output for `ls`, and the `ls` commands are more succinct and readable.

-----

As another example, I'm using this to find the version number of the latest installed Android build tools:
```bash
ls "$sdk_root/build-tools" | sort -r | head -n1
```

ShellCheck would want me to change this to:
```bash
find "$sdk_root/build-tools" -mindepth 1 -maxdepth 1 -printf '%P\n' | sort -r | head -n1
```

... which is equivalent, but the code is much less readable. The `find` command could likely use a comment explaining what it does, whereas the `ls` equivalent could go without (in my opinion at least).

Another issue with the above is that `-printf` is not available in all versions of `find` (e.g. on macOS), requiring more changes to the script, possibly making it even less readable.

P.S.
I know I could rely on `ls` for sorting (or use `tail` instead of `head`), but I wanted to make the examples as similar as possible. I could also be using `sdkmanager` to check the installed build tools, but that also queries external servers (making it much slower, etc.).
